contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,430 @@
+        }
+    }
+
+    function modifyPosition(IPoolManager.PoolKey memory key, IPoolManager.ModifyPositionParams memory params)
+        internal
+        returns (BalanceDelta delta)
+    {
+        // msg.sender is the test contract (aka whoever called addLiquidity/removeLiquidity)
+
+        delta = abi.decode(poolManager.lock(abi.encode(CallbackData(msg.sender, key, params, false))), (BalanceDelta));
+
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance > 0) {
+            CurrencyLibrary.NATIVE.transfer(msg.sender, ethBalance);
+        }
+    }
+
+    function hookModifyPosition(IPoolManager.PoolKey memory key, IPoolManager.ModifyPositionParams memory params)
+        internal
+        returns (BalanceDelta delta)
+    {
+        IERC20Minimal(Currency.unwrap(key.currency0)).approve(address(this), type(uint256).max);
+        IERC20Minimal(Currency.unwrap(key.currency1)).approve(address(this), type(uint256).max);
+
+        delta = abi.decode(poolManager.lock(abi.encode(CallbackData(address(this), key, params, true))), (BalanceDelta));
+
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance > 0) {
+            CurrencyLibrary.NATIVE.transfer(msg.sender, ethBalance);
+        }
+    }
+
+    function lockAcquired(uint256, bytes calldata rawData) external override returns (bytes memory) {
+        require(msg.sender == address(poolManager));
+
+        CallbackData memory data = abi.decode(rawData, (CallbackData));
+
+        BalanceDelta delta = poolManager.modifyPosition(data.key, data.params);
+
+        // check if we are inputting liquidity for token0
+        if (delta.amount0() > 0) {
+            if (data.key.currency0.isNative()) {
+                poolManager.settle{value: uint128(delta.amount0())}(data.key.currency0);
+            } else {
+                IERC20Minimal(Currency.unwrap(data.key.currency0)).transferFrom(
+                    data.sender, address(poolManager), uint128(delta.amount0())
+                );
+                poolManager.settle(data.key.currency0);
+            }
+            // withdrawing liquidity for token0
+        } else {
+            // if withdrawing is because of rebalance
+            if (data.rebalance) {
+                poolManager.take(data.key.currency0, data.sender, uint256(uint128(-delta.amount0())));
+
+                // NOTE: even though we've taken all of the tokens we're owed, we don't set position.tokensOwed to 0
+                // since we need to reinvest into the pool
+                // after reinvestment, we should set the tokens owed to 0
+            } else {
+                poolManager.take(data.key.currency0, data.sender, uint256(uint128(-delta.amount0())));
+
+                // NOTE: commented out because we are never adding the liquidity being taken out to the tokensOwed.
+                // position.tokensOwed0 -= delta.amount0();
+            }
+
+            if (data.key.currency0.isNative()) {
```

Comment:
logic starting at lines 138 and 167 should just be a function that takes a currency and an amount to DRY code. (kind of like balanceOf)

Reply:
also why are we settling down here. if the delta is negative, we're taking....not sure why we have to settle too?

Reply:
addressed in new commit, created functions that take in key and delta and sender address!